packages/shared/src/telemetry.ts

```diff
@@ -0,0 +1,292 @@
+
+    // We will try to grab the SDK data lazily when an event is triggered, so it should always be defined once the event is sent.
+    this.#metadata.sdk = options.sdk!;
+    this.#metadata.sdkVersion = options.sdkVersion!;
+
+    this.#metadata.publishableKey = options.publishableKey ?? '';
+
+    const parsedKey = parsePublishableKey(options.publishableKey);
+    if (parsedKey) {
+      this.#metadata.instanceType = parsedKey.instanceType;
+    }
+
+    if (options.secretKey) {
+      // Only send the first 16 characters of the secret key to to avoid sending the full key. We can still query against the partial key.
+      this.#metadata.secretKey = options.secretKey.substring(0, 16);
+    }
+  }
+
+  get isEnabled(): boolean {
+    if (this.#metadata.instanceType !== 'development') {
+      return false;
+    }
+
+    // In browser or client environments, we most likely pass the disabled option to the collector, but in environments
+    // where environment variables are available we also check for `CLERK_TELEMETRY_DISABLED`.
+    if (this.#config.disabled || (typeof process !== 'undefined' && isTruthy(process.env.CLERK_TELEMETRY_DISABLED))) {
+      return false;
+    }
+
+    return true;
+  }
+
+  get isDebug(): boolean {
+    return this.#config.debug || (typeof process !== 'undefined' && isTruthy(process.env.CLERK_TELEMETRY_DEBUG));
+  }
+
+  record(event: TelemetryEvent['event'], payload: TelemetryEvent['payload']) {
+    const preparedPayload = this.#preparePayload(event, payload);
+
+    this.#logEvent(preparedPayload.event, preparedPayload);
+
+    if (!this.#shouldRecord()) {
+      return;
+    }
+
+    this.#buffer.push(preparedPayload);
+
+    this.#scheduleFlush();
+  }
+
+  #shouldRecord(): boolean {
+    return this.isEnabled && !this.isDebug && Math.random() <= this.#config.samplingRate;
+  }
+
+  #scheduleFlush(): void {
+    // On the server, we want to flush immediately as we have less guarantees about the lifecycle of the process
+    if (typeof window === 'undefined') {
+      this.#flush();
+      return;
+    }
+
+    const isBufferFull = this.#buffer.length >= this.#config.maxBufferSize;
+    if (isBufferFull) {
+      // If the buffer is full, flush immediately to make sure we minimize the chance of event loss.
+      // Cancel any pending flushes as we're going to flush immediately
+      if (this.#pendingFlush) {
+        const cancel = typeof cancelIdleCallback !== 'undefined' ? cancelIdleCallback : clearTimeout;
+        cancel(this.#pendingFlush);
+      }
+      this.#flush();
+      return;
+    }
+
+    // If we have a pending flush, do nothing
+    if (this.#pendingFlush) return;
+
+    if ('requestIdleCallback' in window) {
+      this.#pendingFlush = requestIdleCallback(() => {
+        this.#flush();
+      });
+    } else {
+      // This is not an ideal solution, but it at least waits until the next tick
+      this.#pendingFlush = setTimeout(() => {
+        this.#flush();
+      }, 0);
+    }
+  }
```

Comment:
This was my attempt to implement buffering in a way that mitigates potential data loss. It uses `requestIdleCallback` to schedule sending events during idle time. When an event is recorded, it schedules a flush during idle, and if the buffer limit is reached it immediately flushes the events.

The scheduling of a flesh on every record might be a tad aggressive, but we can tweak later if we want.